### ISSUE TITLE
`systemctl reboot` inside podman machine clears volume mounts

### ISSUE THREAD

**Issue by iamkirkbater** [containers/podman#15976]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

I'm running podman on a new M1 mac, and I need to enable the qemu-user-static package in order to support multiple architectures of packages.  I've noticed that when I `systemctl reboot` from inside the podman machine, as the machine comes back online it does not have my defined mounts anymore.

**Steps to reproduce the issue:**

1. `podman machine init -v $HOME:$HOME`
2. `podman machine ssh`
3. `ls /` - note that you see a `Users` directory
4. `sudo systemctl reboot` 
5. wait for the machine to reboot and `podman machine ssh`
6. `ls /` - note there is no longer a Users directory

**Describe the results you received:**
The mounts defined as part of the init process or in containers.conf are not present when rebooting from within the machine.

**Describe the results you expected:**
I'd expect the mounts to still be present after a reboot.

**Additional information you deem important (e.g. issue happens only occasionally):**
The workaround is simple, you just want to `podman machine stop && podman machine start` but it's still just an extra step when you'd expect this to just come back online with the correct mounts.

**Output of `podman version`:**

```
Client:       Podman Engine
Version:      4.2.1
API Version:  4.2.1
Go Version:   go1.18.6
Built:        Tue Sep  6 15:16:02 2022
OS/Arch:      darwin/arm64

Server:       Podman Engine
Version:      4.2.0
API Version:  4.2.0
Go Version:   go1.18.4
Built:        Thu Aug 11 10:43:11 2022
OS/Arch:      linux/arm64

```

**Output of `podman info`:**

```
host:
  arch: arm64
  buildahVersion: 1.27.0
  cgroupControllers:
  - cpu
  - io
  - memory
  - pids
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: conmon-2.1.4-2.fc36.aarch64
    path: /usr/bin/conmon
    version: 'conmon version 2.1.4, commit: '
  cpuUtilization:
    idlePercent: 96.19
    systemPercent: 3.27
    userPercent: 0.53
  cpus: 1
  distribution:
    distribution: fedora
    variant: coreos
    version: "36"
  eventLogger: journald
  hostname: localhost.localdomain
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 1000000
    uidmap:
    - container_id: 0
      host_id: 501
      size: 1
    - container_id: 1
      host_id: 100000
      size: 1000000
  kernel: 5.19.9-200.fc36.aarch64
  linkmode: dynamic
  logDriver: journald
  memFree: 1717428224
  memTotal: 2051633152
  networkBackend: netavark
  ociRuntime:
    name: crun
    package: crun-1.6-2.fc36.aarch64
    path: /usr/bin/crun
    version: |-
      crun version 1.6
      commit: 18cf2efbb8feb2b2f20e316520e0fd0b6c41ef4d
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +YAJL
  os: linux
  remoteSocket:
    exists: true
    path: /run/user/501/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: true
  serviceIsRemote: true
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.2.0-0.2.beta.0.fc36.aarch64
    version: |-
      slirp4netns version 1.2.0-beta.0
      commit: 477db14a24ff1a3de3a705e51ca2c4c1fe3dda64
      libslirp: 4.6.1
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.5.3
  swapFree: 0
  swapTotal: 0
  uptime: 0h 6m 47.00s
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  volume:
  - local
registries:
  search:
  - docker.io
store:
  configFile: /var/home/core/.config/containers/storage.conf
  containerStore:
    number: 0
    paused: 0
    running: 0
    stopped: 0
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /var/home/core/.local/share/containers/storage
  graphRootAllocated: 106825756672
  graphRootUsed: 2307162112
  graphStatus:
    Backing Filesystem: xfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Using metacopy: "false"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 0
  runRoot: /run/user/501/containers
  volumePath: /var/home/core/.local/share/containers/storage/volumes
version:
  APIVersion: 4.2.0
  Built: 1660228991
  BuiltTime: Thu Aug 11 10:43:11 2022
  GitCommit: ""
  GoVersion: go1.18.4
  Os: linux
  OsArch: linux/arm64
  Version: 4.2.0
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman` or `brew info podman`):**

```
> brew info podman
==> podman: stable 4.2.1 (bottled), HEAD
Tool for managing OCI containers and pods
https://podman.io/
/opt/homebrew/Cellar/podman/4.2.1 (178 files, 48MB) *
  Poured from bottle on 2022-09-27 at 12:02:00
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/podman.rb
License: Apache-2.0
==> Dependencies
Build: go-md2man, go@1.18
Required: qemu
==> Options
--HEAD
	Install HEAD version
==> Caveats
zsh completions have been installed to:
  /opt/homebrew/share/zsh/site-functions
==> Analytics
install: 25,576 (30 days), 63,893 (90 days), 211,139 (365 days)
install-on-request: 24,798 (30 days), 62,485 (90 days), 209,434 (365 days)
build-error: 1 (30 days)

```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/main/troubleshooting.md)**

Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**
This is running on an M1Pro Macbook.

I'm happy to provide any additional details as needed.  Thanks so much for all of your work on this, I'm excited that this is actually a viable alternative to docker now!

**Comment by mheon** [comment]
@baude @ashley-cui PTAL

**Comment by Luap99** [comment]
I think this is the expected behaviour given that we currently manually make the mount calls via ssh after start, see https://github.com/containers/podman/blob/dca5ead2d7ad8ac3b14fed6736c102b571d8baf1/pkg/machine/qemu/machine.go#L661-L690

I you want to fix this you have to mount this inside the VM on boot. I think you could create the proper systemd units via ignition or append lines to fstab. Contributions welcome.

**Comment by iamkirkbater** [comment]
Thanks for pointing me in the right direction.  I'll see what I can do with the time I have to play with this.

**Comment by afbjorklund** [comment]
The implementation assumes "stop" and "start", someone converted the ssh commands to a systemd service which would make it run in boot. Possibly making it harder to mount and umount while running ?

**Comment by iamkirkbater** [comment]
I'm sorry, I'm not quite sure I'm following that last comment.  So the implementation is designed for a physical `podman machine stop && podman machine start` I think I understand that part.  I'm not quite understanding the second part of that.  Do you mean to say that someone deviated from the original intention, or that someone _could_ deviate from the original intention by using the systemd service?

**Comment by afbjorklund** [comment]
It was mostly an implementation detail. I _think_ I abstracted the mount and umount into functions, so the same would be needed with the systemd service. If it is generated by the ignition, then it would be less obvious how to manipulate it at runtime

**Comment by iamkirkbater** [comment]
Fair enough.  I just noticed this happened overnight when I went to go create a container this morning and it failed with the `statfs /Users/me/path/to/file: no such file or directory` error, and when I looked at the tab I was logged into podman machine with I see the following:

```[core@localhost ~]$
Broadcast message from Zincati at Tue 2022-10-04 15:12:24 UTC:
New update 36.20221001.2.0 is available and has been deployed.
If permitted by the update strategy, Zincati will reboot into this update when
all interactive users have logged out, or in 10 minutes, whichever comes
earlier. Please log out of all active sessions in order to let the auto-update
process continue.
[core@localhost ~]$
Broadcast message from Zincati at Tue 2022-10-04 15:21:24 UTC:
New update 36.20221001.2.0 is available and has been deployed.
If permitted by the update strategy, Zincati will reboot into this update when
all interactive users have logged out, or in 1 minute, whichever comes
earlier. Please log out of all active sessions in order to let the auto-update
process continue.
Connection to localhost closed by remote host.```

So I'm hoping I can get to this sooner rather than later, especially if the podman backing machine is going to reboot on it's own which will then cause subsequent container runs that previously worked to not work again.

I'm lucky I saw the error message, because I didn't run the reboot myself this time and I would have been super confused why all of a sudden the mounts I had setup disappeared.

**Comment by iamkirkbater** [comment]
I had a few moments to poke at this last Friday and played around with the idea of just dynamically writing the `/etc/fstab` as part of the `podman machine start` command.  To test this, I manually wrote the `/etc/fstab` file on the machine, and then ran `sudo reboot`.  This worked, the new volumes came back up on reboot!  However, when I attempted to `podman machine stop && podman machine start` again after, the subsequent `podman machine start` command failed because the volumes already existed.

I think what I'll need to do in order to get this working is modify the existing `podman machine init` command to add any mounts to the ignition config file and have the ignition config both create the directory structure the user is looking for as well as the `/etc/fstab` file in order to automatically mount them on startup.  Then I'll have to remove the ssh-mount commands from the start command.

The only thing that worries me about this is that for existing podman-machines, if the user updates the podman remote and then starts their podman machine back up, none of their already-configured mounts will come up.  I'm wondering if there's some kind of annotation that could be added to the machine itself somehow as part of the new init and then we could mount via ssh if that annotation doesn't exist to preserve backwards compatibility.

The next steps for me is learning more about ignition configs and getting that part working first, though, then I'll submit a draft PR with that working and could probably have some further discussion once we get to that point.

**Comment by afbjorklund** [comment]
I think systemd handles `/etc/fstab`, like it handles everything else (soon)

https://www.freedesktop.org/software/systemd/man/systemd.mount.html

**Comment by iamkirkbater** [comment]
I'm hoping to get some time to work on it during my afternoon today, I'm just wondering what the `(soon)` meant?  Is this already being worked on elsewhere?  Or is it still worthwhile for me to try to work on this?  I'd hate to spend a bunch of time trying to implement something if someone else is already working on it.

**Comment by iamkirkbater** [comment]
And of course I couldn't help myself but to power through.  I ended up figuring out how to get past the immutable filesystem stuff using the same commands that are being used with the ssh bits, just they're being run as a prereq to the mount.  Had to add `Restart=on-failure` to the Mount itself because sometimes even though the `Before` directive is on the mkdir service it would still complain about the mountpoint not existing.

Because these live in the `/etc/systemd/system` directory, I believe they _should_ persist through ostree updates.

I have a few todos left on that draft PR, like adding some more unit tests as well as adding some comments.  I think the biggest problem with this will still be deciding how y'all want to help "migrate" existing VMs, I'm not sure what the appropriate path forward is and would be happy to help develop whatever that would be.

If there's any immediate feedback, I'm usually around on Slack.  Thanks!

**Comment by rhatdan** [comment]
@ashley-cui Is this still broken?

**Comment by ashley-cui** [comment]
Still an issue, probably needs some design discussion on dynamic mounts

**Comment by rhatdan** [comment]
I thought someone added a PR to put the mounts in as systemd unit files, which would fire on reboot?

**Comment by ashley-cui** [comment]
Draft PR: https://github.com/containers/podman/pull/16143

**Comment by iamkirkbater** [comment]
Yeah, that was me who started that PR. The problem I'm having with it right now is inconsistency. Sometimes the systemd files all fire in order, sometimes it takes multiple reboots. Once they're all applied they run fine, seemingly forever, but it's not a better UX at all to have to fight the podman machine on your first startup. 

I just haven't had a chance to poke at it again since the last update on it. 

This is my first venture into systemd, so I could just be missing something simple.

**Comment by xiaobitipao** [comment]
Since the mount behavior occurs after the start, as long as the restart after the reboot should solve the problem. Although a bit wordy.

```
podman machine ssh sudo systemctl reboot
podman machine stop
podman machine start
```

**Comment by rhatdan** [comment]
@iamkirkbater are you still working on this?

**Comment by iamkirkbater** [comment]
👋🏼  I haven't had a chance to work on this for a while due to conflicting priorities.  I've just been restarting my podman machine with `podman machine stop && sleep 3 && podman machine start` but haven't had a chance to circle back to this yet.

**Comment by iamkirkbater** [comment]
I also think that I'm at the limits of my `systemd` knowledge with #16143 and I'm not sure if I'd be able to make any more progress with what I have there.

**Comment by rhatdan** [comment]
@Luap99 @ashley-cui @baude Is this still an issue?

**Comment by Luap99** [comment]
yes

**Comment by wnm3** [comment]
Related [issue](https://github.com/containers/podman/issues/19925) for MacOS. Using /etc/synthetic.conf I have a symlink from ~/store -> /store and attempting to mount /store/pipeline with the -v CLI option for the run command fails with the same `Error: statfs /store/pipeline no such file or directory` error **_which keeps the container from running_**.

**Comment by BlaineEXE** [comment]
Does anyone know whether this issue may be related to podman hanging often when my macbook wakes from sleep? I have also been experiencing this mount issue, but I only notice during the few instances of podman not hanging after I step away for a while.

Whenever I open my laptop, any "real" podman command hangs (e.g., `podman images` or `podman ps`), and I have to `podman machine stop && sleep 1 && podman machine start` to get podman to run again.

I have theorized that the podman hanging issue may be related to the mount issue, presuming that the socket mount is failing when podman hangs, but I'm not sure if that's the case. Could anyone suggest what the best ways for me to debug the hanging issue would be?

**Comment by F30** [comment]
This is a really annoying issue given that it also affects automatic updates of the Podman machine. (I assume it's the same bug).

Zincati will apply updates to the machine at random times by default (or at defined times if you configure a [custom update strategy](https://coreos.github.io/zincati/usage/updates-strategy/)). As part of that, the machine will get rebooted and all mounts will be lost, e.g. leading to VS Code dev containers crashing.

**Comment by baude** [comment]
The original bug report should be fixed in Podman 5 with virtio volumes.  I followed the reproducer and the volumes are still useable after the reboot.

**Comment by wnm3** [comment]
Thank you @baude -- I'm still searching my hard drive for containers.conf but I assume I likely added the right commands there so we can close this issue.
